### PR TITLE
Improved Evaluations

### PR DESCRIPTION
**Pull request by farzadab**
This PR does the following:
- Bugfix: AnyInstruct was not using val/train splits (as it had somehow gotten removed from the last PR)
- Faster evaluations by using DDP multiprocessing for text generation
- Validation loss added by default to training

### CODE REVIEW

File: ultravox/training/train.py [hunk 1]
```diff
@@ -170,21 +173,42 @@ def main() -> None:
...
+                    processor=processor,
+                    train_on_inputs=args.train_on_inputs,
+                ),
+                num_samples=num_samples,
+            )
+            for split_args, num_samples in [
+                (data_args, args.num_samples),
+                (val_data_args, args.val_num_samples),
+            ]
+        ]
```
Comment (farzadab) [inline ultravox/training/train.py:203]: This looks ugly, I agree. Let me know if you have a better idea. Other solutions required more copy&pasting.

File: ultravox/training/ddp_utils_test.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import os
+
+import torch.distributed
+from torch import multiprocessing as mp
+
+from ultravox.training import ddp_utils
+
+os.environ["MASTER_ADDR"] = "localhost"
+os.environ["MASTER_PORT"] = "12355"
```
Comment (farzadab) [inline ultravox/training/ddp_utils_test.py:9]: I needed to add some stuff to make it look like I'm running with `torchrun`.

File: ultravox/data/datasets.py [hunk 3]
```diff
@@ -353,16 +353,18 @@ def __init__(self, args: VoiceDatasetArgs) -> None:
...
+        dataset = datasets.load_dataset(
+            "json",
+            "anyinstruct",
+            data_files="https://huggingface.co/datasets/fnlp/AnyInstruct/resolve/main/speech_conv/metadata.jsonl",
+            split="train",
+        ).select(range(NUM_SAMPLES))
+        dataset = dataset.train_test_split(
+            test_size=0.01, seed=args.shuffle_seed, shuffle=True
         )
+        dataset = dataset["train" if args.split == DatasetSplit.TRAIN else "test"]
```
Comment (juberti) [inline ultravox/data/datasets.py:365]: should this just be `dataset[args.split]`?

File: ultravox/training/config_base.py [hunk 4]
```diff
@@ -54,7 +55,7 @@ class TrainConfig:
     optimizer: str = "adamw_torch"
     num_epochs: int = 1
     max_steps: int = 0
-    eval_steps: Optional[int] = None
+    val_steps: Optional[int] = None
```
Comment (juberti) [inline ultravox/training/config_base.py:58]: just checking, did you mean to replace `eval_steps` (as `eval_num_samples` wasn't replaced)

File: ultravox/training/train.py (hunk 1 above)
Comment (juberti) [reply ultravox/training/train.py:203]: maybe just hoist this logic into a `prepare_dataset` function that takes `args` and `num_samples`

**Review by juberti — APPROVED** [review]

File: ultravox/training/config_base.py (hunk 4 above)
Comment (farzadab) [reply ultravox/training/config_base.py:58]: No. The idea is that we have 3 things now: training, validation, final (generative) evaluation.

HF calls the second one `eval` (e.g. `eval_dataset`, `eval_steps`), but I wanted to make sure the names `val` refers to validation and `eval` refers to the final generative evaluation.

On that note, I see that I need to rename `eval_accum_steps` to `val_accum_steps` as well to keep things consistent.

File: ultravox/data/datasets.py (hunk 3 above)
Comment (farzadab) [reply ultravox/data/datasets.py:365]: The issue is we have "validation" and "train" but the `train_test_split` function creates "test" and "train" splits.

**Comment by farzadab** [comment]
Small training run to check for regressions seems to have run fine: https://wandb.ai/fixie/ultravox/runs/q3rh3yxv

File: ultravox/training/train.py (hunk 1 above)
Comment (farzadab) [reply ultravox/training/train.py]: Refactored. The parameter list is a bit long, but I prefer that over either sending the whole train_args or using a nested function.

Comment (juberti) [reply ultravox/training/train.py]: looks great
